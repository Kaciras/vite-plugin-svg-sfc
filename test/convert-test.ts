import { expect, it } from "vitest";
import { convert, resolveFixture, TestOptions } from "./test-helper.ts";
import { SVGSFCConvertor } from "../index.ts";
import { basename } from "path";
import { readFileSync } from "fs";

const strokeSVG = readFileSync(resolveFixture("stroke.svg"), "utf8");

it("should change attributes", async () => {
	expect(await convert("styles-0.svg?sfc")).toMatchSnapshot();
});

it("should change the stroke attribute", async () => {
	expect(await convert("stroke.svg?sfc")).toMatchSnapshot();
});

it("should remove processing instructions in dev", async () => {
	expect(await convert("instruction.svg?sfc", { mode: "development" })).toMatchSnapshot();
});

it("should remove processing instructions in prod", async () => {
	expect(await convert("instruction.svg?sfc")).toMatchSnapshot();
});

it("should extract styles", async () => {
	const code = await convert("styles-0.svg?sfc");
	expect(code.toString()).toMatchSnapshot();
});

it("should not process SVG if svgo option is false", async () => {
	const config = { svgo: false } as const;
	const code = await convert("stroke.svg?sfc", { config });
	expect(code).toBe(`<template>${strokeSVG}</template>`);
});

it("should support configure SVG plugins", () => {
	const convertor = new SVGSFCConvertor({
		svgo: {
			plugins: [
				{ name: "preset-default" },
				{
					name: "modifySVGAttrs",
					params: { foo: 11 },
				},
				{ name: "responsiveSVGAttrs" },
			],
		},
	});
	const resolved = (convertor as any).plugins;
	expect(resolved).toHaveLength(3);
	expect(resolved[0]).toStrictEqual({ name: "preset-default" });
	expect(resolved[1].fn).toBeTypeOf("function");
	expect(resolved[2].fn).toBeTypeOf("function");
});

it("should apply only extractCSS plugin", () => {
	const promise = convert("styles-0.svg?sfc", {
		config: {
			svgo: { plugins: ["extractCSS"] },
		},
	});
	return expect(promise).resolves.toMatchSnapshot();
});

it("should change <svg>'s attributes with svgProps", async () => {
	const config: TestOptions = {
		config: {
			svgProps: {
				":data-foo": "1",	// Add new
				viewBox: "0 0 5 5",	// Replace
			},
		},
	};
	expect(await convert("styles-0.svg?sfc", config)).toMatchSnapshot();
});

it("should change <svg>'s attributes with custom function", async () => {
	const config: TestOptions = {
		config: {
			svgProps(attrs, path, passes) {
				attrs.path = basename(path);
				attrs.passes = passes;
			},
		},
	};
	expect(await convert("styles-0.svg?sfc", config)).toMatchSnapshot();
});
