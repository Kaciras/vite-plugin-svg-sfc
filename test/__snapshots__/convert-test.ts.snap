// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should apply only extractCSS plugin 1`] = `"<template><svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><defs/><rect id="rect" width="10" height="10"/></svg></template><style scoped>#rect { fill: blue; }.st0 { width: 100px; }</style>"`;

exports[`should change <svg>'s attributes with custom function 1`] = `"<template><svg width="1em" height="1em" fill="currentColor" passes="0" path="styles-0.svg" viewBox="0 0 20 20"><path id="rect" d="M0 0h10v10H0z"/></svg></template><style scoped>#rect{fill:#00f}</style>"`;

exports[`should change <svg>'s attributes with svgProps 1`] = `"<template><svg :data-foo="1" width="1em" height="1em" fill="currentColor" viewBox="0 0 5 5"><path id="rect" d="M0 0h10v10H0z"/></svg></template><style scoped>#rect{fill:#00f}</style>"`;

exports[`should change attributes 1`] = `"<template><svg width="1em" height="1em" fill="currentColor" viewBox="0 0 20 20"><path id="rect" d="M0 0h10v10H0z"/></svg></template><style scoped>#rect{fill:#00f}</style>"`;

exports[`should change the stroke attribute 1`] = `"<template><svg id="layer_1" width="1em" height="1em" fill="none" stroke="currentColor" stroke-width="2" data-name="layer 1" viewBox="0 0 92.01 92.55"><path d="m90.02 66.22-31.11-.97-18.43 25.08-8.69-29.89-29.55-9.78 25.74-17.5.17-31.13 24.6 19.07 29.65-9.45-10.54 29.28z"/></svg></template>"`;

exports[`should extract styles 1`] = `"<template><svg width="1em" height="1em" fill="currentColor" viewBox="0 0 20 20"><path id="rect" d="M0 0h10v10H0z"/></svg></template><style scoped>#rect{fill:#00f}</style>"`;

exports[`should remove processing instructions in dev 1`] = `"<template><svg fill="currentColor" height="1em" width="1em"><path d="M0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0z"/></svg></template>"`;

exports[`should remove processing instructions in prod 1`] = `"<template><svg width="1em" height="1em" fill="currentColor"><path d="M0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0z"/></svg></template>"`;
